### ISSUE TITLE
Investigate use of ADL instead of manual delegate declaration

### ISSUE THREAD

**Issue by Nihlus** [GtkSharp/GtkSharp#42]
I'm an avid user of GtkSharp on Linux, Mac, and Windows, and having it support .NET Standard is great. However, from reading through the source, I can see that it's being done through manual delegate declaration and then loading live delegates from function pointers.

Full disclaimer - I am the author of the library in question, so this is a plug, but the library is made for exactly this purpose.

The [ADL][adl] library abstracts this away from the end user, and also provides options for using even faster implementations behind the scenes, while maintaining a flexible and efficient end-user interface.

It would allow a significant reduction in boilerplate code - for example, taking the [`Accel.cs`][accel] class (purely as a quick example). Using ADL, it would be reduced to this:

https://gist.github.com/Nihlus/01f2457406aa5776a7502d371cba6d29

Further reductions could be made by providing ADL with extensions to support marshalling GLib strings and GLib objects down to IntPtrs automatically.

I'd be more than happy to look into this closer, if the maintenance team thinks it could be worthwhile.

[adl]: https://github.com/Firwood-Software/AdvanceDLSupport
[accel]: https://github.com/GtkSharp/GtkSharp/blob/develop/Source/Libs/GtkSharp/Accel.cs

**Comment by cra0zy** [comment]
Don't you need to do a IL rewrite for the performance benefits?

**Comment by Nihlus** [comment]
No, ADL can leverage the faster features without any custom compilation steps, IL weaving, or rewriting. It's the same principle as OpenTK's use of calli, but without the hassle.
